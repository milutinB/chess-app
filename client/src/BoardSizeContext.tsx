import { createContext } from "react";

export const BoardSizeContext: React.Context<number> = createContext(800);