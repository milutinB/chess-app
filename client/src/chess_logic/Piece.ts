import { PlayerColor, BoardCoord, BoardState, PieceSym } from "../types/Types2.d";
var board_config = require("./../board_config.json");
export const RELATIVE_CHESS_ASSEST_DIR: string = "../chess_assets/";

export abstract class Piece {
    constructor(color: PlayerColor, position: BoardCoord) {
        this.position = position;
        this.color = color;
    }

    position: BoardCoord;
    path: string;
    color: PlayerColor;

    abstract getSym() : PieceSym;
    abstract legalMoves() : BoardCoord[];
};

// export class NullPiece extends Piece {
//     constructor(color: PlayerColor, position: BoardCoord) {
//         super(color, position)
//     }

//     legalMoves() : BoardCoord[] {
//         return [];
//     }
// }