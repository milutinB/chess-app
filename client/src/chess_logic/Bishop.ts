import { Piece, RELATIVE_CHESS_ASSEST_DIR } from "./Piece";
import { PlayerColor, BoardCoord, BoardState, RowIndex, ColIndex, PieceSym } from "../types/Types2.d";
var board_config = require("./../board_config.json");

export class Bishop extends Piece {
    constructor(color: PlayerColor, position: BoardCoord) {
        super(color, position);
        this.path = require("../chess_assets/"
        + board_config["pieces"][color]["bishop"]); 
    }

    getSym() : PieceSym {
        return this.color + 'b' as PieceSym;
    }

    legalMoves(): BoardCoord[] {
        let r: number = this.position.r;
        let c: number = this.position.c;
        let rows: number[] = [0, 1, 2, 3, 4, 5, 6, 7];
        let cols: number[] = [0, 1, 2, 3, 4, 5, 6, 7];
        let output: BoardCoord[] = [];
        [1, -1].map((xDir) => {
            [1, -1].map((yDir) => {
                for (var i = 1; i < 8; ++i) {
                    if (rows.includes(r + yDir * i) 
                    && cols.includes(c + xDir * i)) {
                        output.push({
                            r: r + yDir * i as RowIndex,
                            c: c + xDir * i as ColIndex
                        })
                    }
                }
            });
        })
        return output;
    }
}