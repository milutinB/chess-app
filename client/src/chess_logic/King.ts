import { Piece, RELATIVE_CHESS_ASSEST_DIR } from "./Piece";
import { PlayerColor, BoardCoord, BoardState, RowIndex, ColIndex, PieceSym } from "../types/Types2.d";
import board_config from "./../board_config.json";
// var board_config = require("./../board_config.json");

export class King extends Piece {
    constructor(color: PlayerColor, position: BoardCoord) {
        super(color, position);
        this.path = require("../chess_assets/" 
        + board_config["pieces"][color]["king"]); 
    }

    getSym() : PieceSym {
        return this.color + 'k' as PieceSym;
    }

    legalMoves() : BoardCoord[] {
        let r: RowIndex = this.position.r;
        let c: ColIndex = this.position.c;
        let rows: number[] = [0, 1, 2, 3, 4, 5, 6, 7];
        let cols: number[] = [0, 1, 2, 3, 4, 5, 6, 7];
        let output: BoardCoord[] = [];
        [-1, 0, 1].map((dx: number) => {
            [-1, 0, 1].map((dy: number) => {
                let newR: number = r + dy;
                let newC: number = c + dx;
                if (newC != c || newR != r) {
                    if (rows.includes(newC) && cols.includes(newR)) {
                        output.push({
                            r: newR as RowIndex,
                            c: newC as ColIndex
                        })
                    }
                }
            });
        });
        return output;
    }

    hasMoved: boolean = false;
}