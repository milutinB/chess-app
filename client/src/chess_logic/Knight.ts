import { Piece, RELATIVE_CHESS_ASSEST_DIR } from "./Piece";
import { PieceSym, PlayerColor, BoardCoord, BoardState, RowIndex, ColIndex } from "../types/Types2.d";
import board_config from "./../board_config.json";
// var board_config = require("./../board_config.json");

export class Knight extends Piece {
    constructor(color: PlayerColor, position: BoardCoord) {
        super(color, position);
        this.path = require("../chess_assets/"
        + board_config["pieces"][color]["knight"]); 
    }

    getSym() : PieceSym {
        return this.color + 'kn' as PieceSym;
    }

    legalMoves() : BoardCoord[] {
        let r: number = this.position.r;
        let c: number = this.position.c;
        let rows: number[] = [0, 1, 2, 3, 4, 5, 6, 7];
        let cols: number[] = [0, 1, 2, 3, 4, 5, 6, 7];
        let output: BoardCoord[] = [];
        [1, -1].map((xDir: 1 | -1) => {
            [1, -1].map((yDir: 1 | -1) => {
                [1, 2].map((xDist: 1 | 2) => {
                    let yDist: 1 | 2 = xDist == 1 ? 2 : 1;
                    if (rows.includes(r + yDist * yDir)) 
                        if (cols.includes(c + xDist * xDir))
                            output.push({
                                r: r + yDist * yDir as RowIndex,
                                c: c + xDist * xDir as ColIndex
                        });
                })
            })
        })
        return output;
    }
}