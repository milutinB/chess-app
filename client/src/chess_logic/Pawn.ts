import { Piece, RELATIVE_CHESS_ASSEST_DIR } from "./Piece";
import { PieceSym, PlayerColor, BoardCoord, BoardState, RowIndex, ColIndex } from "../types/Types2.d";
import board_config from "./../board_config.json";
// var board_config = require("./../board_config.json");


export class Pawn extends Piece {
    constructor(color: PlayerColor, position: BoardCoord) {
        super(color, position);
        this.path = require("../chess_assets/"
        + board_config["pieces"][color]["pawn"]); 
    }

    getSym() : PieceSym {
        return this.color + 'p' as PieceSym;
    }

    legalMoves() : BoardCoord[] {
        let r: RowIndex = this.position.r;
        let c: ColIndex = this.position.c;
        let rows: number[] = [0, 1, 2, 3, 4, 5, 6, 7];
        let cols: number[] = [0, 1, 2, 3, 4, 5, 6, 7];
        let output: BoardCoord[] = [];
        let yDir: number = this.color == 'w' ? 1 : -1;


        if (rows.includes(r + yDir)) {
            output.push({
                r: r + yDir as RowIndex,
                c: c
            });
            if (cols.includes(c + 1))
                output.push({
                    r: r + yDir as RowIndex,
                    c: c + 1 as ColIndex
                })
            if (cols.includes(c - 1))
                output.push({
                    r: r + yDir as RowIndex,
                    c: c - 1 as ColIndex
                })
        }
        if ((this.color == 'w' && this.position.r == 1) ||
            (this.color == 'b' && this.position.r == 6))
            output.push({
                r: r + 2 * yDir as RowIndex,
                c: c
            });
        
        return output;
    }
}