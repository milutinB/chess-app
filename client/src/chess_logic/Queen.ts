import { Piece, RELATIVE_CHESS_ASSEST_DIR } from "./Piece";
import { PieceSym, PlayerColor, BoardCoord, BoardState, RowIndex, ColIndex } from "../types/Types2.d";
var board_config = require("./../board_config.json");


export class Queen extends Piece {
    constructor(color: PlayerColor, position: BoardCoord) {
        super(color, position);
        this.path = require("../chess_assets/" 
        + board_config["pieces"][color]["queen"]); 
    }
    
    getSym() : PieceSym {
        return this.color + 'q' as PieceSym;
    }

    legalMoves() : BoardCoord[] {
        let r: RowIndex = this.position.r;
        let c: ColIndex = this.position.c;
        let rows: number[] = [0, 1, 2, 3, 4, 5, 6, 7];
        let cols: number[] = [0, 1, 2, 3, 4, 5, 6, 7];
        let output: BoardCoord[] = [];
        [1, -1].map((xDir) => {
            [1, -1].map((yDir) => {
                for (var i = 1; i < 8; ++i) {
                    if (rows.includes(r + yDir * i) 
                    && cols.includes(c + xDir * i)) {
                        output.push({
                            r: r + yDir * i as RowIndex,
                            c: c + xDir * i as ColIndex
                        })
                    }
                }
            });
        })
        for (var i = 0; i < 8; ++i) {
            if (i != r)
                output.push({r: i as RowIndex, c: c});
            if (i != c)
                output.push({r: r, c: i as ColIndex});
        }
        return output;
    }
}