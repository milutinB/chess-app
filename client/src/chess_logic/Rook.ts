import { Piece, RELATIVE_CHESS_ASSEST_DIR } from "./Piece";
import { PieceSym, PlayerColor, BoardCoord, BoardState, RowIndex, ColIndex } from "../types/Types2.d";
var board_config = require("./../board_config.json");


export class Rook extends Piece {
    constructor(color: PlayerColor, position: BoardCoord) {
        super(color, position);
        this.path = require("../chess_assets/" 
        + board_config["pieces"][color]["rook"]); 
    }

    getSym() : PieceSym {
        return this.color + 'r' as PieceSym;
    }

    legalMoves() : BoardCoord[] {
        let r: RowIndex = this.position.r;
        let c: ColIndex = this.position.c;
        let output: BoardCoord[] = [];

        for (var i = 0; i < 8; ++i) {
            if (i != r)
                output.push({r: i as RowIndex, c: c});
            if (i != c)
                output.push({r: r, c: i as ColIndex});
        }
        return output;
    }

    hasMoved: boolean = false;
}