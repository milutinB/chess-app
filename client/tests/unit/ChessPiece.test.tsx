import ChessPiece from "../../src/components/ChessPiece";
import { BoardCoord, ColIndex, PlayerColor, RowIndex, SquareData } from "../../src/types/Types2.d";
import { Piece } from "../../src/chess_logic/Piece";
import { Knight } from "../../src/chess_logic/Knight";
import { King } from "../../src/chess_logic/King";
import { Queen } from "../../src/chess_logic/Queen";
import { Pawn } from "../../src/chess_logic/Pawn";
import { Bishop } from "../../src/chess_logic/Bishop";
import { Rook } from "../../src/chess_logic/Rook";
import { BoardSizeContext } from "../../src/BoardSizeContext";
import * as renderer from 'react-test-renderer';
import { render, cleanup, screen } from '@testing-library/react';

afterEach(cleanup);

let boardSize = 800;

it('renders correctly', () => {
    let pawn: SquareData = new Pawn('w', {r: 1, c: 0});
    const tree: any = renderer.create(
        <BoardSizeContext.Provider value={boardSize}>
            <ChessPiece piece={pawn}/>
        </BoardSizeContext.Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

let wKnightPath: string = "Chess_nlt45.svg"
let wRookPath: string = "Chess_rlt45.svg";
let wBishopPath: string = "Chess_blt45.svg";
let wPawnPath: string = "Chess_plt45.svg"
let wKingPath: string = "Chess_klt45.svg";
let wQueenPath: string = "Chess_qlt45.svg";
let bKnightPath: string = "Chess_ndt45.svg";
let bRookPath: string = "Chess_rdt45.svg";
let bBishopPath: string = "Chess_bdt45.svg";
let bPawnPath: string = "Chess_pdt45.svg";
let bKingPath: string = "Chess_kdt45.svg";
let bQueenPath: string = "Chess_qdt45.svg";


describe('Piece Component', () => {
    it('renders white pawn correctly', () => {
        render(
            <BoardSizeContext.Provider value={boardSize}>
                    <ChessPiece piece={new Pawn('w', {r: 1, c: 0})}/>
            </BoardSizeContext.Provider>
        );

        expect(screen.getByRole('piece-image').getAttribute("href"))
        .toEqual(wPawnPath);
    });

    it('renders white rook correctly', () => {
        render(
            <BoardSizeContext.Provider value={boardSize}>
                    <ChessPiece piece={new Rook('w', {r: 1, c: 0})}/>
            </BoardSizeContext.Provider>
        );

        expect(screen.getByRole('piece-image').getAttribute("href"))
        .toEqual(wRookPath);
    });

    it('renders white bishop correctly', () => {
        render(
            <BoardSizeContext.Provider value={boardSize}>
                    <ChessPiece piece={new Bishop('w', {r: 1, c: 0})}/>
            </BoardSizeContext.Provider>
        );

        expect(screen.getByRole('piece-image').getAttribute("href"))
        .toEqual(wBishopPath);
    });

    it('renders white Knight correctly', () => {
        render(
            <BoardSizeContext.Provider value={boardSize}>
                    <ChessPiece piece={new Knight('w', {r: 1, c: 0})}/>
            </BoardSizeContext.Provider>
        );

        expect(screen.getByRole('piece-image').getAttribute("href"))
        .toEqual(wKnightPath);
    });

    it('renders white king correctly', () => {
        render(
            <BoardSizeContext.Provider value={boardSize}>
                    <ChessPiece piece={new King('w', {r: 1, c: 0})}/>
            </BoardSizeContext.Provider>
        );

        expect(screen.getByRole('piece-image').getAttribute("href"))
        .toEqual(wKingPath);
    });

    it('renders white Queen correctly', () => {
        render(
            <BoardSizeContext.Provider value={boardSize}>
                    <ChessPiece piece={new Queen('w', {r: 1, c: 0})}/>
            </BoardSizeContext.Provider>
        );

        expect(screen.getByRole('piece-image').getAttribute("href"))
        .toEqual(wQueenPath);
    });

    it('renders black pawn correctly', () => {
        render(
            <BoardSizeContext.Provider value={boardSize}>
                    <ChessPiece piece={new Pawn('b', {r: 1, c: 0})}/>
            </BoardSizeContext.Provider>
        );

        expect(screen.getByRole('piece-image').getAttribute("href"))
        .toEqual(bPawnPath);
    });

    it('renders black rook correctly', () => {
        render(
            <BoardSizeContext.Provider value={boardSize}>
                    <ChessPiece piece={new Rook('b', {r: 1, c: 0})}/>
            </BoardSizeContext.Provider>
        );

        expect(screen.getByRole('piece-image').getAttribute("href"))
        .toEqual(bRookPath);
    });

    it('renders black bishop correctly', () => {
        render(
            <BoardSizeContext.Provider value={boardSize}>
                    <ChessPiece piece={new Bishop('b', {r: 1, c: 0})}/>
            </BoardSizeContext.Provider>
        );

        expect(screen.getByRole('piece-image').getAttribute("href"))
        .toEqual(bBishopPath);
    });

    it('renders black Knight correctly', () => {
        render(
            <BoardSizeContext.Provider value={boardSize}>
                    <ChessPiece piece={new Knight('b', {r: 1, c: 0})}/>
            </BoardSizeContext.Provider>
        );

        expect(screen.getByRole('piece-image').getAttribute("href"))
        .toEqual(bKnightPath);
    });

    it('renders black king correctly', () => {
        render(
            <BoardSizeContext.Provider value={boardSize}>
                    <ChessPiece piece={new King('b', {r: 1, c: 0})}/>
            </BoardSizeContext.Provider>
        );

        expect(screen.getByRole('piece-image').getAttribute("href"))
        .toEqual(bKingPath);
    });

    it('renders black Queen correctly', () => {
        render(
            <BoardSizeContext.Provider value={boardSize}>
                    <ChessPiece piece={new Queen('b', {r: 1, c: 0})}/>
            </BoardSizeContext.Provider>
        );

        expect(screen.getByRole('piece-image').getAttribute("href"))
        .toEqual(bQueenPath);
    });
});
