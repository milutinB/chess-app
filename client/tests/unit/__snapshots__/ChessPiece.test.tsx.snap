// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<svg>
  <image
    height={100}
    href="Chess_plt45.svg"
    role="piece-image"
    width={100}
  />
</svg>
`;
